plutus-core/plutus-ir/src/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -373,13 +372,13 @@ withVarsOfBinding r (DatatypeBind _ dt) k = do
     (_tyconstrDef, constrDefs, destrDef) <- compileDatatypeDefs r (original dt)
     -- ignore the generated rhs terms of constructors/destructor
     let structorDecls = PIR.defVar <$> destrDef:constrDefs
-    -- normalize, then rename, then only introduce the vardecl to scope
     foldr normRenameScope k structorDecls
     where
+      -- normalize, then introduce the vardecl to scope
       normRenameScope :: VarDecl TyName Name uni fun (Provenance a)
                       -> TypeCheckM uni fun c e res -> TypeCheckM uni fun c e res
       normRenameScope v acc = do
-          normRenamedTy <- rename =<< (normalizeTypeM $ _varDeclType v)
+          normRenamedTy <- normalizeTypeM $ _varDeclType v
           withVar (_varDeclName v) (void <$> normRenamedTy) acc
```

Comment:
1. `normalizeTypeM` requires and preserves global uniqueness, i.e. `v` has to satisfy global uniqueness and that's enough to produce  `normRenamedTy` satisfying global uniqueness
2. `withVar` takes care to rename the looked up value, otherwise even an explicit call to `rename` wouldn't suffice, 'cause one could use the same type variable twice and looking it up would break global uniqueness

Here's what we do for lambdas:

```
inferTypeM (LamAbs ann n dom body)  = do
    checkKindM ann dom $ Type ()
    vDom <- normalizeTypeM $ void dom
    TyFun () <<$>> pure vDom <<*>> withVar n vDom (inferTypeM body)
```

It's the same `normalizeTypeM`-then-`withVar` logic, no renaming required.